File: bindings/c/src/group_box.cpp
// Copyright (c) 2025 Daniel T. McGinnis
// SPDX-License-Identifier: MIT

#include <nux/group_box.h>
#include "handles.hpp"
#include "../include/nux/core.h"

NUXC_EXTERNC
nux_group_box_t*
nux_group_box_create (
    nux_window_element_t* parent
) {
    auto element{nux_impl_map(parent)};

    if (
        auto enclosing_layout{dynamic_cast<nux::layout*>(element)};
        enclosing_layout
    ) {
        auto result{
            new nux_group_box_t{
                nux::group_box{*enclosing_layout},
                nux::string{}
            }
        };
        return result;
    }

    if (
        auto enclosing_group_box{dynamic_cast<nux::group_box*>(element)};
        enclosing_group_box
    ) {
        auto result{
            new nux_group_box_t{
                nux::group_box{
                    enclosing_group_box->grab_native_handle(),
                    enclosing_group_box->grab_content_layout()
                },
                nux::string{}
            }
        };
        return result;
    }

    return nullptr;
}

NUXC_EXTERNC
void
nux_group_box_destroy (
    nux_group_box_t* group_box
) {
    delete group_box;
}

NUXC_EXTERNC
nux_widget_t*
nux_group_box_upcast (
    nux_group_box_t* group_box
) {
    return nux_impl_map(static_cast<nux::widget*>(nux_impl_map(group_box)));
}

NUXC_EXTERNC
nux_layout_t*
nux_group_box_grab_content_layout (
    nux_group_box_t* group_box
) {
    return nux_impl_map(group_box->handle.grab_content_layout());
}

NUXC_EXTERNC
const nuxchar_t*
nux_group_box_get_title (
    nux_group_box_t* group_box,
    int* size
) {
    group_box->title = group_box->handle.get_title();
    if (size) {
        *size = group_box->title.size();
    }
    return reinterpret_cast<const nuxchar_t*>(group_box->title.data());
}

NUXC_EXTERNC
void
nux_group_box_set_title (
    nux_group_box_t* group_box,
    const nuxchar_t* new_title
) {
    group_box->handle.set_title(
        reinterpret_cast<nux::string_view::const_pointer>(new_title)
    );
}

NUXC_EXTERNC
void
nux_group_box_add_widget (
    nux_group_box_t* group_box,
    nux_widget_t* child
) {
    auto& content_layout_handle{nux_impl_map(group_box)->grab_content_layout()};
    auto child_handle{nux_impl_map(child)};
    content_layout_handle.grab_children().push_back(
        std::unique_ptr<nux::widget>{child_handle}
    );
    content_layout_handle.update();
}

File: bindings/rs/nux/src/group_box.rs
// Copyright (c) 2025 Daniel T. McGinnis
// SPDX-License-Identifier: MIT

use std::ffi::*;
use std::rc::*;
use std::cell::*;
use crate::bindings::*;
use crate::window_element::*;
use crate::layout::*;
use crate::widget::*;

#[derive(Clone)]
pub struct GroupBox {
    widget_handle_ref: WidgetHandleReference
}

impl GroupBox {
    pub fn grab_content_layout (
        &self
    ) -> Layout {
        return Layout::from_raw(
            unsafe {
                nux_group_box_grab_content_layout(
                    self.widget_handle_ref.borrow().handle.native_handle
                        as *mut nux_group_box_t
                ) as *mut nux_layout_t
            }
        );
    }

    pub fn get_title (
        &self
    ) -> String {
        unsafe {
            let mut size: std::os::raw::c_int = 0;
            let data: *const nuxchar_t = nux_group_box_get_title(
                self.widget_handle_ref.borrow().handle.native_handle
                    as *mut nux_group_box_t,
                &mut size as *mut std::os::raw::c_int
            );
            return str::from_utf8(
                std::slice::from_raw_parts(
                    data as *const u8,
                    size as usize
                )
            ).unwrap().to_string();
        }
    }

    pub fn set_title (
        &self,
        title: &str
    ) {
        unsafe {
            nux_group_box_set_title(
                self.widget_handle_ref.borrow().handle.native_handle
                    as *mut nux_group_box_t,
                CString::new(title).unwrap().as_ptr()
                    as *const u8
            );
        };
    }

    pub fn add_widget<T> (
        &self
    ) -> T
    where T: SimpleWidget
    {
        unsafe {
            let widget = T::new(&self.widget_handle_ref.borrow().handle);
            nux_group_box_add_widget(
                self.widget_handle_ref.borrow().handle.native_handle
                    as *mut nux_group_box_t,
                widget.grab_widget_handle().borrow().handle.native_handle
                    as *mut nux_widget_t
            );
            return widget;
        };
    }
}

impl WindowElement for GroupBox {
    fn grab_handle (
        &self
    ) -> WindowElementHandle {
        return self.grab_widget_handle().borrow().handle;
    }
}

impl Widget for GroupBox {
    fn grab_widget_handle (
        &self
    ) -> WidgetHandleReference {
        return self.widget_handle_ref.clone();
    }
}

impl SimpleWidget for GroupBox {
    fn new (
        parent: &WindowElementHandle
    ) -> Self {
        return Self {
            widget_handle_ref: Rc::new(
                RefCell::new(
                    WidgetHandle {
                        handle: WindowElementHandle {
                            native_handle: unsafe {
                                nux_group_box_create(
                                    parent.native_handle
                                ) as *mut nux_window_element_t
                            }
                        },
                        kind: NativeWidgetKind::GroupBox,
                        graphics_box_render_event_handler: None,
                        push_button_trigger_event_handler: None,
                        check_box_toggle_event_handler: None,
                        radio_button_toggle_event_handler: None
                    }
                )
            )
        };
    }
}